presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -164,8 +164,44 @@ public PlanNode visitPlan(PlanNode node, RewriteContext<Expression> context)
         @Override
         public PlanNode visitProject(ProjectNode node, RewriteContext<Expression> context)
         {
-            Expression inlinedPredicate = ExpressionTreeRewriter.rewriteWith(new ExpressionSymbolInliner(node.getAssignments()), context.get());
-            return context.defaultRewrite(node, inlinedPredicate);
+            Map<Symbol, Expression> assignments = node.getAssignments();
+
+            Set<Symbol> nonDeterministicSymbols = assignments.entrySet().stream()
+                    .filter(s -> !DeterminismEvaluator.isDeterministic(s.getValue()))
+                    .map(Map.Entry::getKey)
+                    .collect(Collectors.toSet());
+
+            Expression deterministicPredicate;
+            List<Expression> nonDeterministicConjuncts = new ArrayList<>();
+
+            if (nonDeterministicSymbols.isEmpty()) {
```

Comment:
This is a special case of the logic in the else branch, no? If so, just get rid to the `if ... else` and do what's in the `else` branch unconditionally.